### PR TITLE
Adding connection stats handler

### PR DESCRIPTION
**Pull request by nsivabalan**
Adding connection stats handler to track open connections and disconnections.

Tested manually using jconsole that all 3 new metrics get updated. 

SLA: 5 mins
Reviewers: Casey, Ming

### CODE REVIEW

File: ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
+/**
+ * Copyright 2015 LinkedIn Corp. All rights reserved.
```
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java]: this now 2016

File: ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Handler that tracks connection establishment statistics.
+ */
+class ConnectionStatsHandler extends ChannelDuplexHandler {
+  private final NettyMetrics metrics;
+  private AtomicLong openConnections;
```
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java]: `final`

File: ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+
+  public ConnectionStatsHandler(NettyMetrics metrics) {
+    this.metrics = metrics;
+    openConnections = new AtomicLong(0);
+    metrics.registerConnectionsStatsHandler(openConnections);
+  }
+
+  /**
+   * Calls {@link ChannelHandlerContext#connect(java.net.SocketAddress, java.net.SocketAddress, io.netty.channel.ChannelPromise)} to forward
```
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java]: change doc.

File: ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+   * to the next {@link io.netty.channel.ChannelOutboundHandler} in the {@link io.netty.channel.ChannelPipeline}.
+   *
+   * Sub-classes may override this method to change behavior.
+   */
+  @Override
+  public void connect(ChannelHandlerContext ctx, SocketAddress remoteAddress, SocketAddress localAddress,
+      ChannelPromise future)
+      throws Exception {
```
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java]: does not throw `Exception`

File: ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    logger.trace("Channel Connected " + remoteAddress);
+    metrics.connectionsConnectedCount.inc();
+    openConnections.incrementAndGet();
+    super.connect(ctx, remoteAddress, localAddress, future);
+  }
+
+  /**
+   * Calls {@link ChannelHandlerContext#disconnect(ChannelPromise)} to forward
```
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java]: change doc

File: ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+   * to the next {@link io.netty.channel.ChannelOutboundHandler} in the {@link io.netty.channel.ChannelPipeline}.
+   *
+   * Sub-classes may override this method to change behavior.
+   */
+  @Override
+  public void disconnect(ChannelHandlerContext ctx, ChannelPromise future)
+      throws Exception {
```
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java]: does not throw `Exception`.

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyMetrics.java [hunk 7]
```diff
@@ -26,6 +28,8 @@
  */
 class NettyMetrics {
 
+  MetricRegistry metricRegistry;
```
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyMetrics.java]: `private final`.

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyMetrics.java [hunk 8]
```diff
@@ -107,6 +111,11 @@
   // NettyServer
   public final Histogram nettyServerShutdownTimeInMs;
   public final Histogram nettyServerStartTimeInMs;
+  // ConnectionStatsHandler
+  Counter connectionsConnectedCount;
```
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyMetrics.java]: `public final` (the `public` is not really required but keep it for consistency)

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyMetrics.java [hunk 9]
```diff
@@ -107,6 +111,11 @@
...
+  Counter connectionsDisconnectedCount;
+  private Gauge<Long> openConnections;
```
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyMetrics.java]: don't need to declare the `Gauge` as a class member.

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyMetrics.java [hunk 10]
```diff
@@ -237,5 +247,24 @@ public NettyMetrics(MetricRegistry metricRegistry) {
...
     nettyServerStartTimeInMs = metricRegistry.histogram(MetricRegistry.name(NettyServer.class, "StartTimeInMs"));
+    // ConnectionStatsHandler
+    connectionsConnectedCount =
+        metricRegistry.counter(MetricRegistry.name(NettyServer.class, "ConnectionsConnectedCount"));
+    connectionsDisconnectedCount =
+        metricRegistry.counter(MetricRegistry.name(NettyServer.class, "ConnectionsDisconnectedCount"));
+  }
+
+  /**
+   * Register {@link ConnectionStatsHandler#openConnections} to track open connections
```
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyMetrics.java]: `openConnections` is a private variable. Don't make it a part of a `public` (or package private) function's doc.

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyMetrics.java [hunk 11]
```diff
@@ -237,5 +247,24 @@ public NettyMetrics(MetricRegistry metricRegistry) {
...
+   * @param openConnectionsCount open connections count to be tracked
+   */
+  public void registerConnectionsStatsHandler(AtomicLong openConnectionsCount) {
+    openConnections = new Gauge<Long>() {
+      @Override
+      public Long getValue() {
+        return openConnectionsCount.get();
+      }
+    };
+    metricRegistry.register(MetricRegistry.name(NettyServer.class, "OpenConnections"), openConnections);
```
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyMetrics.java]: The owner should be the  one who actually owns the metric which is `ConnectionStatsHandler`.

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyMetrics.java [hunk 12]
```diff
@@ -237,5 +247,24 @@ public NettyMetrics(MetricRegistry metricRegistry) {
     // NettyServer
     nettyServerShutdownTimeInMs = metricRegistry.histogram(MetricRegistry.name(NettyServer.class, "ShutdownTimeInMs"));
     nettyServerStartTimeInMs = metricRegistry.histogram(MetricRegistry.name(NettyServer.class, "StartTimeInMs"));
+    // ConnectionStatsHandler
+    connectionsConnectedCount =
```
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyMetrics.java]: owner of both is `ConnectionStatsHandler`.

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyServer.java [hunk 13]
```diff
@@ -62,16 +63,19 @@
    * @param publicAccessLogger the {@link PublicAccessLogger} that can be used for public access logging
    * @param restServerState the {@link RestServerState} that can be used to check the health of the system
    *                              to respond to health check requests
+   * @param connectionStatsHandler the {@link ConnectionStatsHandler} to collect metrics related to connections
    */
   public NettyServer(NettyConfig nettyConfig, NettyMetrics nettyMetrics, RestRequestHandler requestHandler,
-      PublicAccessLogger publicAccessLogger, RestServerState restServerState) {
+      PublicAccessLogger publicAccessLogger, RestServerState restServerState,
+      ConnectionStatsHandler connectionStatsHandler) {
```
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyServer.java]: why is this passed in? None of the other handlers are passed in.

File: ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+import java.net.SocketAddress;
+import java.util.concurrent.atomic.AtomicLong;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Handler that tracks connection establishment statistics.
+ */
+class ConnectionStatsHandler extends ChannelDuplexHandler {
```
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java]: you can add the decorator `@Shareable` here.

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyServer.java [hunk 15]
```diff
@@ -91,8 +95,10 @@ public void start()
         public void initChannel(SocketChannel ch)
             throws Exception {
           ch.pipeline()
-              // for http encoding/decoding. Note that we get content in 8KB chunks and a change to that number has
-              // to go here.
+              // connection stats handler
```
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyServer.java]: would be better to give a one line description of what the handler does.

File: ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java (hunk 4 above)
Comment (nsivabalan) [inline ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java]: super.connect() throws Exception

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyMetrics.java (hunk 12 above)
Comment (nsivabalan) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyMetrics.java]: Why do you think ConnectionStatsHandler should own this? In that case, we can also argue for health check metrics, public access log metrics and so on. I feel we can have all the metrics in one place (NettyMetrics)

File: ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java (hunk 14 above)
Comment (nsivabalan) [inline ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java]: Can you explain why we need that. I don't see any other classes appended with this annotation in our code base.

Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java]: http://netty.io/4.0/api/io/netty/channel/ChannelHandler.Sharable.html

You use the annotation when you don't want a new instance of the class created for every connection (which is the default). In essence, you are saying that the same object can be shared across multiple connections without creating a race condition (same as what you are current code is doing now).

Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java]: This condition is not true for any of the other handlers.

Comment (nsivabalan) [inline ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java]: thanks. Updated

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyMetrics.java (hunk 12 above)
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyMetrics.java]: Discussed this offline - just need to change the `NettyServer` into `ConnectionStatsHandler`

**Comment by vgkholla** [comment]
LGTM except for that one comment.

Assuming everything is formatted and built.

**Comment by nsivabalan** [comment]
@vgkholla : Addressed all comments. Yes, built and formatted

**Comment by vgkholla** [comment]
LGTM

**Comment by vgkholla** [comment]
@cgtz @xiahome You guys are the primary reviewers on this. Could you take a look and merge if this looks good?

**Comment by cgtz** [comment]
LGTM

**Comment by xiahome** [comment]
Since there is some issue in this PR, will you make a quick fix or temporarily disable this PR?

**Comment by vgkholla** [comment]
@nsivabalan Please pull in the recent changes and update the PR.

**Comment by vgkholla** [comment]
@nsivabalan Please pull in the recent changes and update the PR. That should resolve the problems with this PR.

**Comment by vgkholla** [comment]
@nsivabalan Could you try rebasing and give this a shot. This should be simple enough and we should get this PR merged since it has been around for a while.

**Comment by xiahome** [comment]
@vgkholla I remember the initial issues of this PR are: 1) not clear if `connect()` and `disconnect()` are triggered by corresponding event as expected; 2) it caused failure for a mocking test. Could you explain why recent changes help this PR fix the issue? I will come to you tomorrow.

**Comment by vgkholla** [comment]
The second point you mention was not because of this patch (and it was not a mocking test) and was fixed in #307. For your first point, I thought we agreed that @nsivabalan would do a simple test by examining the JMX metrics to make sure that the metrics are fired (it is even possible to check this programmatically).

**Comment by nsivabalan** [comment]
I will work on this later today and will update.

**Comment by nsivabalan** [comment]
@xiahome @cgtz : Patch is good to be reviewed now

File: ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+import java.util.concurrent.atomic.AtomicLong;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Handler that tracks connection establishment statistics.
+ */
+@ChannelHandler.Sharable
+class ConnectionStatsHandler extends ChannelDuplexHandler {
```
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java]: needs to be `public` like the other handlers.

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyMetrics.java [hunk 17]
```diff
@@ -108,7 +112,11 @@
   // NettyServer
   public final Histogram nettyServerShutdownTimeInMs;
   public final Histogram nettyServerStartTimeInMs;
-  // PublicAccessLogHandler
+  // ConnectionStatsHandler
+  public final Counter connectionsConnectedCount;
+  public final Counter connectionsDisconnectedCount;
+
+  // PublicAccessLogRequestHandler
```
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyMetrics.java]: Why this change? This is called `PublicAccessLogHandler` now.

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyMetrics.java [hunk 18]
```diff
@@ -240,5 +249,24 @@ public NettyMetrics(MetricRegistry metricRegistry) {
...
+        metricRegistry.counter(MetricRegistry.name(ConnectionStatsHandler.class, "ConnectionsConnectedCount"));
+    connectionsDisconnectedCount =
+        metricRegistry.counter(MetricRegistry.name(ConnectionStatsHandler.class, "ConnectionsDisconnectedCount"));
+  }
+
+  /**
+   * Registers the {@link ConnectionStatsHandler} to track open connections
+   * @param openConnectionsCount open connections count to be tracked
+   */
+  public void registerConnectionsStatsHandler(AtomicLong openConnectionsCount) {
```
Comment (vgkholla) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyMetrics.java]: doesn't need to be `public` I think.

**Comment by vgkholla** [comment]
minor comments. LGTM.

**Comment by nsivabalan** [comment]
Addressed the comments from gopal

File: ambry-rest/src/test/java/com.github.ambry.rest/NettyServerTest.java [hunk 19]
```diff
@@ -125,6 +125,8 @@ private NettyServer getNettyServer(Properties properties)
       @Override
       protected void initChannel(SocketChannel ch) {
         ch.pipeline()
+            // connection stats handler to track connection related metrics
+            .addLast("ConnectionStatsHandler", new ConnectionStatsHandler(nettyMetrics))
```
Comment (vgkholla) [inline ambry-rest/src/test/java/com.github.ambry.rest/NettyServerTest.java]: `ConnectionStatsHandler.getInstance(nettyMetrics)`

**Comment by nsivabalan** [comment]
updated the patch

File: ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+import java.util.concurrent.atomic.AtomicLong;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Handler that tracks connection establishment statistics.
+ */
+@ChannelHandler.Sharable
+public class ConnectionStatsHandler extends ChannelDuplexHandler {
```
Comment (xiahome) [inline ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java]: My original question still holds, which is why a `ChannelDuplexHandler` is needed. I changed it to `ChannelInboundHandlerAdapter` and all the tests still passed, which at least means `channelActive` and `channelInactive` are supplied correctly.

File: ambry-rest/src/main/java/com.github.ambry.rest/NettyMetrics.java [hunk 21]
```diff
@@ -240,5 +249,24 @@ public NettyMetrics(MetricRegistry metricRegistry) {
...
+        metricRegistry.counter(MetricRegistry.name(ConnectionStatsHandler.class, "ConnectionsConnectedCount"));
+    connectionsDisconnectedCount =
+        metricRegistry.counter(MetricRegistry.name(ConnectionStatsHandler.class, "ConnectionsDisconnectedCount"));
+  }
+
+  /**
+   * Registers the {@link ConnectionStatsHandler} to track open connections
+   * @param openConnectionsCount open connections count to be tracked
+   */
+  void registerConnectionsStatsHandler(AtomicLong openConnectionsCount) {
```
Comment (xiahome) [inline ambry-rest/src/main/java/com.github.ambry.rest/NettyMetrics.java]: `final AutomicLong openConnectionCount`

**Comment by xiahome** [comment]
@nsivabalan Have you done any test to ensure that `channelActive` and `channelInactive` can be correctly called?

File: ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java (hunk 20 above)
Comment (xiahome) [inline ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java]: And I guess the two methods are also triggered by right events, hope this could be verified by some test.

Comment (nsivabalan) [inline ambry-rest/src/main/java/com.github.ambry.rest/ConnectionStatsHandler.java]: fixed it

**Comment by xiahome** [comment]
Siva tested using local deployment and correctness of metrics is verified. The code mostly look good to me except for the minor comments above.

**Comment by xiahome** [comment]
LGTM.
